test/UniswapV2PairTest.sol

```diff
@@ -104,38 +140,84 @@ contract UniswapV2PairTest is Test {
+        return token1Amount - divCeil(kAdjusted, newBalance0Adjusted);
+    }
+
     function testOptimisticTestCase1() public {
-        runOptimisticSwapTestCase(997000000000000000, expandTo18Decimals(5), expandTo18Decimals(10), expandTo18Decimals(1));
+        runOptimisticSwapTestCase(997000000000000000, expandTo18Decimals(5), expandTo18Decimals(10));
     }
 
     function testOptimisticTestCase2() public {
-        runOptimisticSwapTestCase(997000000000000000, expandTo18Decimals(10), expandTo18Decimals(5), expandTo18Decimals(1));
+        runOptimisticSwapTestCase(997000000000000000, expandTo18Decimals(10), expandTo18Decimals(5));
     }
 
     function testOptimisticTestCase3() public {
-        runOptimisticSwapTestCase(997000000000000000, expandTo18Decimals(5), expandTo18Decimals(5), expandTo18Decimals(1));
+        runOptimisticSwapTestCase(997000000000000000, expandTo18Decimals(5), expandTo18Decimals(5));
     }
 
     function testOptimisticTestCase4() public {
-        runOptimisticSwapTestCase(expandTo18Decimals(1), expandTo18Decimals(5), expandTo18Decimals(5), 1003009027081243732);
+        runOptimisticSwapTestCase(expandTo18Decimals(1), expandTo18Decimals(5), expandTo18Decimals(5));
+    }
+
+    function testOptimisticFuzzTestCase(uint outputVal, uint token0Val, uint token1Val) public {
+        (uint outputAmount, uint token0Amount, uint token1Amount) = generateOptimisticFuzzCase(outputVal, token0Val, token1Val);
+        runOptimisticSwapTestCase(outputAmount, token0Amount, token1Amount);
+    }
+
+    function generateOptimisticFuzzCase(uint outputAmountSeed, uint token0AmountSeed, uint token1AmountSeed) private
+            returns (uint _outputAmount, uint _token0Amount, uint _token1Amount) {
+
+        // Token 0 needs at least: A) one output amount to take out and B) one remaining, otherwise all removals will fail:
+        uint minToken0Amount = 2;
+        // Token 0 needs to leave at least 2 in the global supply for an input transfer that will be one or greater with fees (beyond one) taken out:
+        uint maxToken0Amount = MAXIMUM_SUPPLY - 2;
+        _token0Amount = mapSeedToRange(token0AmountSeed, minToken0Amount, maxToken0Amount);
+
+        // The liquidity between the token amounts needs to be 1001 or greater or else mint will fail:
+        uint minToken1Amount = divCeil((MINIMUM_LIQUIDITY + 1) ** 2, _token0Amount);
+        _token1Amount = mapSeedToRange(token1AmountSeed, minToken1Amount, MAXIMUM_SUPPLY);
+
+        // We need to make sure there is enough liquidity to have one remaining, so max liquidity-based output is always one less than token amount:
+        uint maxOutputAmountLiquidity = _token0Amount - 1;
+        // We must also ensure enough tokens remain in global supply to allow the input transfer (when incorporating the 0.03% fee, rounding down):
+        uint maxOutputAmountTotalSupply = ((MAXIMUM_SUPPLY - _token0Amount) * 997) / 1000;
```

Comment:
Makes sure that adding back fees (dividing output amount by 99.7%) does not force the input transfer to exceed remaining global supply not moved into the pair.